---
title: "Differential_expression"
author: "Erika Mehrhoff"
date: "2023-07-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#if (!requireNamespace("BiocManager", quietly = TRUE))
#install.packages("BiocManager")
#BiocManager::install("DESeq2")

library(tidyverse)
library(DESeq2)
```

Read in info about samples (sample name and condition and sex)
```{r}
# pull in csv (like excel file) and name it df (dataframe)
df <- read.csv(file= "data/sample_sheet.csv", header=T) #50

# make these variables factors (will explain later)
df$exercise <- as.factor(df$exercise)
df$alcohol <- as.factor(df$alcohol)
df$sex <- as.factor(df$sex)

# split by sex
df_m <- filter(df, df$sex=="male") # 28
df_f <- filter(df, df$sex=="female") # 28

# counting up number of animals in each condition by separating the data by conditions and sex (just to double check)
df_m_sed_con <- filter(df_m, exercise=="sedentary" & alcohol=="control") #
df_f_sed_con <- filter(df_f, exercise=="sedentary" & alcohol=="control") #
df_m_sed_alc <- filter(df_m, exercise=="sedentary" & alcohol=="binge") #
df_f_sed_alc <- filter(df_f, exercise=="sedentary" & alcohol=="binge") # 
df_m_exe_alc <- filter(df_m, exercise=="exercise" & alcohol=="binge") #
df_f_exe_alc <- filter(df_f, exercise=="exercise" & alcohol=="binge") #
df_m_exe_con <- filter(df_m, exercise=="exercise" & alcohol=="control") #
df_f_exe_con <- filter(df_f, exercise=="exercise" & alcohol=="control") #
df_m_cage <- filter(df_m, exercise=="cage_control") #
df_f_cage <- filter(df_f, exercise=="cage_control") #
```

Read in RSEM data and make count matrix
```{r}
# Read in csv file with names of samples
samples = read.table(file= "data/samples.csv", header=T) # need a file that lists the name of the samples to be able to run the for loop

# for loop to pull in the .genes.results file for each sample and then merge them together into one big dataframe called "cnts"
for(i in samples$sample){ # column from samples file we just pulled in that lists the samples to run the for loop through 
  x = read.table(file=paste("data/genes_results/",i, ".genes.results",sep=""),sep="\t",header=TRUE) # read.table is a command to actually pull in the .genes.results files 
  x = x[,c("gene_id", "transcript_id.s.", "expected_count")] # telling it what columns we want to keep from the .genes.results
  colnames(x)[3] = i
  if(i!=samples$sample[1]) cnts = merge(x,cnts,by=c("gene_id","transcript_id.s."))
  if(i==samples$sample[1]) cnts = x # using this to merge each file that comes in with the previous ones that were pulled in 
}
```

Check order of count matrix and sample sheet
```{r}
all(colnames(cnts[,-2])== df$ID)

# reorder columns or rows so they match
```

Read in GTF file (same one used for gusing assembly in the pipeline)
```{r}
# file path to downloaded gtf (need to have the gtf somewhere on your local machine)

gtf <- read.table(file=)

# filter on transcripts
transcripts <- 

anno <-

```

Filter Transcripts based on read counts to prep for DESeq
```{r}
# many filtering methods, no consensus- some a lot more conservative 

# liberal
# Deseq wants round numbers so we are going to round them here:
cnts <- round(cnts)
# now we want to get rid of all genes with 0 across all samples.
# too many 0 will bias normalization 
filtered <- cnts[rowSums(cnts) > 1,]


# more conservative
filtered <- cnts[rowSums(cnts[,-2]>20)>((ncol(cnts)-2)*0.75),] # 50% instead (<100,000)
```

More DESeq prep
```{r}
# samplesheet condition needs to be factored for DESeq

```

Create a DESeq dataset
```{r}
# install.packages("DESeq2")
library(DESeq2)

#dds DESeq Data Set
# full model
dds <- DESeqDataSetFromMatrix(countData = counts_filtered,
                              # this is our counts data we read in
                              colData = df,
                              # telling DeSeq what env variable to use for sample sheet
                              design = ~ exercise + alcohol + sex + exercise:alcohol + exercise:sex + alcohol:sex)
                              # perhaps most important is "condition" is a factor in

# 1st reduced model (looking at exercise)
dds2 <- DESeq(dds,test="LRT",reduced= ~ alcohol + sex + exercise:alcohol + exercise:sex + alcohol:sex ,fitType="local")
results(dds2)
exercise <- results(dds2)
#want stat, padj, and pvalue (log2 fine if 2 groups 2 levels)


# 2nd reduced (looking at alcohol)
#dds3 <- DESeq(dds,test="LRT",reduced= ~ exercise + sex + exercise:alcohol + exercise:sex + alcohol:sex ,fitType="local")
#results(dds3)


# etc...

```

Consolidate and Summary Stats
```{r}
results1 <- as.data.frame(exercise)[,c("stat","pvalue","padj")]


```

Ensembl 
```{r}

```

