---
title: "New_DE_Theo"
author: "Theo Andre"
date: "2023-11-29"
output: html_document
---
BN7!

Fastq files were downloaded from Novogene onto Alpine at CU Boulder.
Files location: /pl/active/IBG/erika/alc_exer_rna/

All analyses are located in: /projects/erme3555/alc_ex_rna/

MD5 checksum values were compared to ensure files were downloaded intact. 
In the folder with the fastq files "md5sum -c MD5.txt" was run and the output read "OK" for all samples.

The raw read counts were determined for each fastq file. The counts are located in /projects/erme3555/alc_ex_rna/rawReadCounts.txt 

Nextflow workflow system and the rnaseq nf-core pipeline were used to process the fastq files and return count data.
Full details are in the MultiQC file, including the packages and versions that were used. File location: /projects/erme3555/alc_ex_rna/results/multiqc/star_rsem/multiqc_data/

The ".genes.results" files were downloaded to my local machine and are in the project folder and are imported in this script.

QC data can be found in:
/projects/erme3555/alc_ex_rna/results3/multiqc/star_rsem/
/projects/erme3555/alc_ex_rna/results3/fastqc/

BigWig files for each sample were downloaded to my local machine and were imported into IGV for visualization.

Package Setup
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# BiocManager is used to install DESeq2
#if (!requireNamespace("BiocManager", quietly = TRUE))
#install.packages("BiocManager")
#BiocManager::install("DESeq2")

# load necessary packages
library(tidyverse)
library(DESeq2)
library(RColorBrewer)
library(pheatmap)
library(biomaRt)
```

Read in info about samples (sample name and condition and sex)
```{r}
# pull in csv (like excel file) and name it df (dataframe)
df <- read.csv(file= "data/sample_sheet.csv", header=T) #50

# make these variables factors (will explain later)
df$exercise <- as.factor(df$exercise)
df$alcohol <- as.factor(df$alcohol)
df$sex <- as.factor(df$sex)

# split by sex
df_m <- filter(df, df$sex=="male") # 28
df_f <- filter(df, df$sex=="female") # 28

# counting up number of animals in each condition by separating the data by conditions and sex (just to double check)
df_m_sed_con <- filter(df_m, exercise=="sedentary" & alcohol=="control") # 6
df_f_sed_con <- filter(df_f, exercise=="sedentary" & alcohol=="control") # 6
df_m_sed_alc <- filter(df_m, exercise=="sedentary" & alcohol=="binge") # 5
df_f_sed_alc <- filter(df_f, exercise=="sedentary" & alcohol=="binge") # 5
df_m_exe_alc <- filter(df_m, exercise=="exercise" & alcohol=="binge") # 5
df_f_exe_alc <- filter(df_f, exercise=="exercise" & alcohol=="binge") # 5
df_m_exe_con <- filter(df_m, exercise=="exercise" & alcohol=="control") # 6
df_f_exe_con <- filter(df_f, exercise=="exercise" & alcohol=="control") # 6
df_m_cage <- filter(df_m, exercise=="cage_control") # 6
df_f_cage <- filter(df_f, exercise=="cage_control") # 6

# quick look at df
head(df, 5)
```
There are 56 samples with 28 females and 28 males. There are 3 factor variables: Exercise (exercise and sedentary), Alcohol (binge and control), and Sex (male and female). There are 5-6 samples in every grouping of exercise, alcohol, and sex.

Raw Read Counts
```{r}
# import txt file created on server 
reads <- read.table("data/rawReadCounts.txt")

# clean up dataframe
reads <- reads[,c(1,3)]
names(reads)[names(reads) == 'V3'] <- 'read_count'
names(reads)[names(reads) == 'V1'] <- 'sample'
reads$sample<-gsub("/pl/active/IBG/erika/alc_exer_rna/01.RawData/","",as.character(reads$sample))

# look at spread of reads
min(reads$read_count)
max(reads$read_count)
mean(reads$read_count)
hist(reads$read_count)
```


Read in RSEM data and make count matrix
```{r}
# remove unnecessary environment objects
rm(df_f, df_m, df_f_cage, df_m_cage, df_m_sed_con, df_f_sed_con, df_m_sed_alc, df_f_sed_alc, df_m_exe_alc, df_f_exe_alc, df_m_exe_con, df_f_exe_con)

# Read in csv file with names of samples
samples = read.table(file= "data/samples.csv", header=T) # need a file that lists the name of the samples to be able to run the for loop

#filenames <- list.files(path="./data/genes_results/")
#for (i in 1:length(filenames)){
 #x <- tximport(paste0("./data/genes_results/",filenames[i]), type = "rsem", txIn = TRUE, txOut = TRUE)
 # y <- as.data.frame(x)
#y <- mutate(y, id = rownames(y))
#y <- dplyr::select(y, counts, id)
#y$counts <- as.numeric(y$counts)
#colnames(y) <- c(filenames[i], "transcript_id") 
#if(i==1) counts=y
#if(i!=1) counts <- merge(y, counts, by= "transcript_id")
#}
merge_trans <- read.table("data/new_genes_results/rsem.merged.transcript_counts.tsv", header=TRUE)
merge_gene <- read.table("data/new_genes_results/rsem.merged.gene_counts.tsv", header=TRUE)

# for loop to pull in the .genes.results file for each sample and then merge them together into one big dataframe called "cnts"
for(i in samples$sample){ # column from samples file we just pulled in that lists the samples to run the for loop through 
  x = read.table(file=paste("data/new_genes_results/",i, ".genes.results",sep=""),sep="\t",header=TRUE) # read.table is a command to actually pull in the .genes.results files 
  x = x[,c("gene_id", "transcript_id.s.", "expected_count")] # telling it what columns we want to keep from the .genes.results
  colnames(x)[3] = i
  if(i!=samples$sample[1]) cnts = merge(x,cnts,by=c("gene_id","transcript_id.s."))
  if(i==samples$sample[1]) cnts = x # using this to merge each file that comes in with the previous ones that were pulled in 
}



# quick look at count data
head(cnts, 5)
```
The counts matrix ("cnts") has 32,285 rows of genes/transcripts. The first 2 columns of the dataframe label the gene and transcript. 

```{r}
sum(str_count(cnts$transcript_id.s., pattern = ","))
```
Transcripts combined within rows leads to the loss of 7040 transcript counts.


Check order of count matrix and sample sheet
```{r}
# For DESeq we need the samples to be in the same order within the count matrix ("cnts") and the samplesheet ("df")
all(colnames(cnts[,3:ncol(cnts)])== df$ID) # if TRUE you're good, if FALSE have to reorder

```


Filter Transcripts based on read counts to prep for DESeq
```{r}
# many filtering methods, no consensus- some a lot more conservative 

# liberal
# Deseq wants counts to be integers
columns <- c(3:ncol(cnts))
cnts[,columns] <- lapply(columns, function(x) as.integer(cnts[[x]]))

# now we want to get rid of all genes with 0 across all samples.
# too many 0 will bias normalization 
filtered <- cnts[rowSums(cnts[,c(3:ncol(cnts))]) > 1,]
# 19,594

# more conservative
#filtered <- cnts[rowSums(cnts[,-2]>20)>((ncol(cnts)-2)*0.75),] 
# 12,280

```
The read counts were filtered by getting rid of any genes/transcripts that didn't have a single read across all samples. This a liberal filtering method. After filterting there are 19,594 genes/transcripts left.


DESeq2- create data set and full model
```{r}
# The count matrix has a column for gene ids and transcript ids, however mutliple transcirpts aligned to some genes and therefore the transcripts column has commas and can't be used.Trying with gene ids
# The count data and col data need to have equal rows/columns so make gene id be the row names
rownames(filtered) = filtered$gene_id
filtered <- filtered[,c(3:ncol(filtered))]

# For now get rid of cage controls because will mess up the groups
df_cage <- droplevels(filter(df, exercise!="cage_control"))
keep <- df_cage$ID
filt_cage <- filtered[,keep]

# recheck order matches
all(colnames(filt_cage)== df_cage$ID)

# save final count data as csv and as r data
write.csv(filt_cage, "data/final_counts.csv", row.names=TRUE)
write_rds(filt_cage, "data/final_counts.rds")

#dds DESeq Data Set
# full model
dds <- DESeqDataSetFromMatrix(countData = filt_cage,
                              # this is our counts data we read in
                              colData = df_cage,
                              # telling DeSeq what env variable to use for sample sheet
                              design = ~ exercise + alcohol + sex + exercise:alcohol + exercise:sex + alcohol:sex + alcohol:exercise:sex)
                              # most important is "condition" is a factor 
```
The cage controls add in a third factor level for exercise and alcohol, which complicates DESeq2. They were removed for the time being, leaving 44 samples. Likely we'll compare all animals to the cage controls.

Visualize
```{r}
# Heatmap of the sample-to-sample distances
sampleDists <- dist(t(assay(dds)))

sampleDistMatrix <- as.matrix(sampleDists)
#rownames(sampleDistMatrix) <- paste(vsd$strain, vsd$sex, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

# cluster dendrogram
plot(hclust(as.dist(1-cor(filt_cage)))) # not on normalized counts

# Principal component plot of the samples
pcaData <- DESeq2::plotPCA(vst(dds), intgroup=c("exercise", "alcohol", "sex"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=exercise)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
ggplot(pcaData, aes(PC1, PC2, color=alcohol)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
ggplot(pcaData, aes(PC1, PC2, color=sex)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

DESeq2- Main effects
```{r}
# exercise
dds1 <- DESeq(dds,test="LRT",reduced= ~ alcohol + sex + alcohol:sex,fitType="local")
exerc <- results(dds1)

# alcohol
dds2 <- DESeq(dds,test="LRT",reduced= ~ exercise + sex + exercise:sex ,fitType="local")
alc <- results(dds2)

# sex
dds3 <- DESeq(dds,test="LRT",reduced= ~ exercise + alcohol + exercise:alcohol ,fitType="local")
sex <- results(dds3)
```

DESeq2- 2 way interactions
```{r}
# exercise x alcohol
dds4 <- DESeq(dds,test="LRT",reduced= ~ exercise + alcohol + sex + exercise:sex + alcohol:sex ,fitType="local")
exerc_alc <- results(dds4)

# exercise x sex
dds5 <- DESeq(dds,test="LRT",reduced= ~ exercise + alcohol + sex + exercise:alcohol + alcohol:sex,fitType="local")
exerc_sex <- results(dds5)

# alcohol x sex
dds6 <- DESeq(dds,test="LRT",reduced= ~ exercise + alcohol + sex + exercise:alcohol + exercise:sex,fitType="local")
alc_sex <- results(dds6)
```

DESeq2- 3 way interaction
```{r}
# exercise x alcohol x sex
dds7 <- DESeq(dds,test="LRT",reduced= ~ exercise + alcohol + sex + exercise:alcohol + exercise:sex + alcohol:sex,fitType="local")
exerc_alc_sex <- results(dds7)
```

Try without sex to see if more DEGs
```{r}
df_cage_ex <- df_cage[,-4]
dds_ex <- DESeqDataSetFromMatrix(countData = filt_cage,
                              # this is our counts data we read in
                              colData = df_cage_ex,
                              # telling DeSeq what env variable to use for sample sheet
                              design = ~ exercise + alcohol + exercise:alcohol)

# exercise
dds_ex_r <- DESeq(dds_ex,test="LRT",reduced= ~ alcohol,fitType="local")
exerc_ex <- results(dds_ex_r)

# alcohol
dds_ex_r_a <- DESeq(dds_ex,test="LRT",reduced= ~ exercise,fitType="local")
exerc_ex_a <- results(dds_ex_r_a)

# exercise x alcohol
dds_ex_r_in <- DESeq(dds_ex,test="LRT",reduced= ~ exercise + alcohol,fitType="local")
exerc_ex_in <- results(dds_ex_r_in)

results_ex_1 <- as.data.frame(exerc_ex)[,c("stat","pvalue","padj")]
results_ex_2 <- as.data.frame(exerc_ex_a)[,c("stat","pvalue","padj")]
results_ex_3 <- as.data.frame(exerc_ex_in)[,c("stat","pvalue","padj")]

nrow(filter(results_ex_1, padj<.05)) # 92
nrow(filter(results_ex_2, padj<.05)) # 7
nrow(filter(results_ex_3, padj<.05)) # 0

nrow(filter(results_ex_1, padj<.1)) # 199
nrow(filter(results_ex_2, padj<.1)) # 9
nrow(filter(results_ex_3, padj<.1)) # 0

nrow(filter(results_ex_1, padj<.15)) # 319
nrow(filter(results_ex_2, padj<.15)) # 25
nrow(filter(results_ex_3, padj<.15)) # 3

# FDR
nrow(filter(results_ex_1, pvalue<.0005)) # 117
nrow(filter(results_ex_2, pvalue<.0005)) # 44
nrow(filter(results_ex_3, pvalue<.0005)) # 24
```

Consolidate and Summary Stats
```{r}
results1 <- as.data.frame(exerc)[,c("stat","pvalue","padj")]
results2 <- as.data.frame(alc)[,c("stat","pvalue","padj")]
results3 <- as.data.frame(sex)[,c("stat","pvalue","padj")]
results4 <- as.data.frame(exerc_alc)[,c("stat","pvalue","padj")]
results5 <- as.data.frame(exerc_sex)[,c("stat","pvalue","padj")]
results6 <- as.data.frame(alc_sex)[,c("stat","pvalue","padj")]
results7 <- as.data.frame(exerc_alc_sex)[,c("stat","pvalue","padj")]

# number of significant DEGs
nrow(filter(results1, padj<.05)) # 6
nrow(filter(results2, padj<.05)) # 1
nrow(filter(results3, padj<.05)) # 7
nrow(filter(results4, padj<.05)) # 0
nrow(filter(results5, padj<.05)) # 0
nrow(filter(results6, padj<.05)) # 0
nrow(filter(results7, padj<.05)) # 0

# number of significant DEGs at p adjusted .1
nrow(filter(results1, padj<.1)) # 40
nrow(filter(results2, padj<.1)) # 13
nrow(filter(results3, padj<.1)) # 9
nrow(filter(results4, padj<.1)) # 0
nrow(filter(results5, padj<.1)) # 0
nrow(filter(results6, padj<.1)) # 0
nrow(filter(results7, padj<.1)) # 0


colnames(results1) <- paste("exercise_",colnames(results1),sep="")
colnames(results2) <- paste("alcohol_",colnames(results2),sep="")
colnames(results3) <- paste("sex_",colnames(results3),sep="")
colnames(results4) <- paste("exercise_alcohol_",colnames(results4),sep="")
colnames(results5) <- paste("exercise_sex_",colnames(results5),sep="")
colnames(results6) <- paste("alcohol_sex_",colnames(results6),sep="")
colnames(results7) <- paste("exercise_alcohol_sex",colnames(results7),sep="")
# all 19594


combined_results <- merge(results1,results2,by=0)
combined_results <- merge(combined_results,results3,by.x="Row.names",by.y=0)
combined_results <- merge(combined_results,results4,by.x="Row.names",by.y=0)
combined_results <- merge(combined_results,results5,by.x="Row.names",by.y=0)
combined_results <- merge(combined_results,results6,by.x="Row.names",by.y=0)
combined_results <- merge(combined_results,results7,by.x="Row.names",by.y=0)

names(combined_results)[names(combined_results) == 'Row.names'] <- 'gene_id'

```

Ensembl to annotate
```{r}
# Read in GTF file (same one used for guiding assembly in the pipeline)
# file path to downloaded gtf (need to have the gtf somewhere on your local machine)

gtf <- read.table(file="Rattus_norvegicus.mRatBN7.2.110.gtf", fill =TRUE)

# filter on transcripts and genes
#transcript <- gtf[gtf$V3=="transcript",] # 54,993
gene <- gtf[gtf$V3=="gene",] # 30,562

# prep datasets
#names(transcript)[names(transcript) == 'V22'] <- 'gene_name'
names(gene)[names(gene) == 'V16'] <- 'gene_name'

#transcript$V21 <- as.factor(transcript$V21)
gene$V15 <- as.factor(gene$V15)

#transcript <- filter(transcript, V21 == "gene_name") # 49412
gene <- filter(gene, V15 == "gene_name") # 26, 195

# pull in ensembl info
ensembl <- useEnsembl(biomart="ensembl",dataset = "rnorvegicus_gene_ensembl")
get_anno <- getBM(attributes=c("ensembl_gene_id",
                              "ensembl_transcript_id",
                              "external_gene_name",
                              "description",
                              "gene_biotype"),
                 values = gene$gene_name, # might want to try transcript df
                 mart = ensembl)

# check there's no duplicates 
all(get_anno==get_anno[!duplicated(get_anno),])

# merge gtf and ensembl info
names(get_anno)[names(get_anno) == 'ensembl_gene_id'] <- 'gene_id'
names(gene)[names(gene) == 'V10'] <- 'gene_id'
ensembl_merge <- merge(get_anno, gene, by="gene_id") # 49412

ensembl_gene <- ensembl_merge %>%
  group_by(gene_name) %>%
  summarize(gene_symbol = paste(unique(external_gene_name), collapse="; "),
            description = paste(unique(description), collapse="; "),
            ensembl_gene_id = paste(unique(gene_id), collapse="; "),
            biotype = paste(unique(gene_biotype), collapse="; ")) # reformat and get rid of duplicates 
# 25,882

gene <- gene[,c(10, 16, 22)] # 26,195
ensembl_gene <- ensembl_gene[, c(1, 3:5)] # 25,882
names(ensembl_gene)[names(ensembl_gene) == 'ensembl_gene_id'] <- 'gene_id'

#gtf_info <- merge(gene,ensembl_gene, by="gene_name", all.x=TRUE) # put gtf transcripts together with ensembl info 
gtf_info <- merge(gene,ensembl_gene, by="gene_id", all.x=TRUE)

save(gtf_info, file="data/ensembl_info.Rdata")
```

Merge DESeq results and annotation and prep for GO analysis
```{r}
# merge DESeq results and annotation info 
combined_results <- merge(gtf_info,combined_results, by="gene_id", all=TRUE) 
all(combined_results$gene_name.x==combined_results$gene_name.y, na.rm = TRUE)
combined_results <- combined_results[,-4]
names(combined_results)[names(combined_results) == 'gene_name.x'] <- 'gene_name'
names(combined_results)[names(combined_results) == 'V22'] <- 'gene_type'
combined_results <- combined_results[,-5]
write.csv(combined_results, "data/final_results_BN7_genelevel.csv", row.names=TRUE)
# 28176

# For GO analysis
all_genes <- combined_results[,c(1:2)]
all_genes <- data.frame(gene_id= all_genes)
write.csv(all_genes, "data/all_genes.csv", row.names=TRUE)
# 29,017

# Did with .05 but could switch to .1
# exercise
res_1 <- combined_results[, c("gene_id", "gene_id","exercise_padj")]
res_1 <- filter(res_1, exercise_padj<.05)
res_1 <- res_1[, c("gene_id")]
res_1 <- data.frame(gene_id= res_1)
write.csv(res_1, "data/exercise_genes.csv", row.names=TRUE)

# alcohol
res_2 <- combined_results[, c("gene_id", "gene_id","alcohol_padj")]
res_2 <- filter(res_2, alcohol_padj<.05)
res_2 <- res_2[, c("gene_id")]
res_2 <- data.frame(gene_id= res_2)
write.csv(res_2, "data/alcohol_genes.csv", row.names=TRUE)

# sex
res_3 <- combined_results[, c("gene_id","gene_id", "sex_padj")]
res_3 <- filter(res_3, sex_padj<.05)
res_3 <- res_3[, c("gene_id")]
res_3 <- data.frame(gene_id= res_3)
write.csv(res_3, "data/sex_genes.csv", row.names=TRUE)

# exercise x alcohol
res_4 <- combined_results[, c("gene_id","gene_id", "exercise_alcohol_padj")]
res_4 <- filter(res_4, exercise_alcohol_padj<.05)
res_4 <- res_4[, c("gene_id")]
res_4 <- data.frame(gene_id= res_4)
write.csv(res_4, "data/exercise_alcohol_genes.csv", row.names=TRUE)

# exercise x sex
res_5 <- combined_results[, c("gene_id","gene_id", "exercise_sex_padj")]
res_5 <- filter(res_5, exercise_sex_padj<.05)
res_5 <- res_5[, c("gene_id")]
res_5 <- data.frame(gene_id= res_5)
write.csv(res_5, "data/exercise_sex_genes.csv", row.names=TRUE)

# alcohol x sex
res_6 <- combined_results[, c("gene_id","gene_id", "alcohol_sex_padj")]
res_6 <- filter(res_6, alcohol_sex_padj<.05)
res_6 <- res_6[, c("gene_id")]
res_6 <- data.frame(gene_id= res_6)
write.csv(res_6, "data/alcohol_sex_genes.csv", row.names=TRUE)

# exercise x alcohol x sex
res_7 <- combined_results[, c("gene_id","gene_id", "exercise_alcohol_sexpadj")]
res_7 <- filter(res_7, exercise_alcohol_sexpadj<.05)
res_7 <- res_7[, c("gene_id")]
res_7 <- data.frame(gene_id= res_7)
write.csv(res_7, "data/exercise_alc_sex_genes.csv", row.names=TRUE)

```
Have list of significant DEGs for each comparison of interest and upload along with background gene expression to PANTHER.

Volcano plots- need representation of fold change
```{r}
ggplot(df, aes(x = log2FoldChange, y = -log10(padj), color = -log10(df$padj) > 90)) + 
  geom_point()
# saving volcano plot
ggsave("data/volcano.pdf")
```

WCGNA
```{r}

```
